### PR TITLE
Add better error handling for OpenSSL failures

### LINKED ISSUE #16101: Test failure with OpenSSL 3.0
### Description

FYI

```
There were 4 errors:

1) Cake\Test\TestCase\Http\Client\Auth\OauthTest::testRsaSigningWithPassphraseString
openssl_sign(): Supplied key param cannot be coerced into a private key

/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/src/Http/Client/Auth/Oauth.php:224
/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/src/Http/Client/Auth/Oauth.php:72
/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/tests/TestCase/Http/Client/Auth/OauthTest.php:469

2) Cake\Test\TestCase\Http\Client\Auth\OauthTest::testRsaSigningStringWithPassphraseString
openssl_sign(): Supplied key param cannot be coerced into a private key

/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/src/Http/Client/Auth/Oauth.php:224
/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/src/Http/Client/Auth/Oauth.php:72
/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/tests/TestCase/Http/Client/Auth/OauthTest.php:505

3) Cake\Test\TestCase\Http\Client\Auth\OauthTest::testRsaSigningWithPassphraseFile
openssl_sign(): Supplied key param cannot be coerced into a private key

/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/src/Http/Client/Auth/Oauth.php:224
/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/src/Http/Client/Auth/Oauth.php:72
/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/tests/TestCase/Http/Client/Auth/OauthTest.php:543

4) Cake\Test\TestCase\Http\Client\Auth\OauthTest::testRsaSigningStringWithPassphraseFile
openssl_sign(): Supplied key param cannot be coerced into a private key

/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/src/Http/Client/Auth/Oauth.php:224
/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/src/Http/Client/Auth/Oauth.php:72
/builddir/build/BUILD/cakephp-6b1cf0e74bbfbb73ecdb1e8a5d0fd36230930931/tests/TestCase/Http/Client/Auth/OauthTest.php:583

```

Notice: upcoming Fedora 36 will have OpenSSL 3.0 and upcoming RHEL 9 (CentOS, Alma, Rocky...)

### CakePHP Version

4.3.1

### PHP Version

_No response_

### PR DESCRIPTION
**Pull request by markstory**
Raise errors when OpenSSL cannot read the private key file.

Refs #16101

### CODE REVIEW

File: tests/test_app/config/key_with_passphrase.pem [hunk 1]
```diff
@@ -1,18 +1,42 @@
 -----BEGIN RSA PRIVATE KEY-----
 Proc-Type: 4,ENCRYPTED
-DEK-Info: DES-CBC,E65DB7AE7A05EF23
+DEK-Info: AES-128-CBC,E634910BEDC73C123DE64ACFF88B339C
```
Comment (markstory) [inline tests/test_app/config/key_with_passphrase.pem:3]: The fedora failures were caused by this key using unsupported formats.

File: tests/TestCase/Http/Client/Auth/OauthTest.php [hunk 2]
```diff
@@ -583,12 +541,16 @@ public function testRsaSigningStringWithPassphraseFile(): void
...
-        );
+        $this->assertSignatureFormat($result);
         $expected = 0;
         $this->assertSame($expected, ftell($passphrase));
     }
+
+    protected function assertSignatureFormat($result)
+    {
+        $this->assertMatchesRegularExpression(
+            '/oauth_signature="[a-zA-Z0-9\/=+]+"/',
```
Comment (markstory) [inline tests/TestCase/Http/Client/Auth/OauthTest.php:575]: I don't think there is much value in asserting the precise signed value. Knowing that there is a signed value is good enough.

File: tests/TestCase/Http/Client/Auth/OauthTest.php [hunk 3]
```diff
@@ -25,41 +25,27 @@
  */
 class OauthTest extends TestCase
 {
-    private $privateKeyString = '-----BEGIN RSA PRIVATE KEY-----
```
Comment (markstory) [inline tests/TestCase/Http/Client/Auth/OauthTest.php:28]: These were duplicated from the files.

File: src/Http/Client/Auth/Oauth.php [hunk 4]
```diff
@@ -220,6 +220,9 @@ protected function _rsaSha1(Request $request, array $credentials): string
             $credentials['privateKeyPassphrase'] = $passphrase;
         }
         $privateKey = openssl_pkey_get_private($credentials['privateKey'], $credentials['privateKeyPassphrase']);
+        if ($privateKey === false) {
+            throw new RuntimeException('Could not open privateKey file. Got ' . openssl_error_string());
```
Comment (othercorey) [inline src/Http/Client/Auth/Oauth.php]: openssl_error_string() is a queue so we might need to pull all the errors to get the last if something else failed first.

Can we just write a helper function and call it if any of these openssl apis fail?

Comment (markstory) [reply src/Http/Client/Auth/Oauth.php]: > Can we just write a helper function and call it if any of these openssl apis fail?

We could. We could have something like `checkOpenSslError()` that raises on an error.

File: src/Http/Client/Auth/Oauth.php [hunk 5]
```diff
@@ -367,4 +371,21 @@ protected function _encode(string $value): string
...
+    /**
+     * Check for SSL errors and raise if one is encountered.
+     *
+     * @return void
+     */
+    protected function checkSslError(): void
+    {
+        $error = '';
+        while ($text = openssl_error_string()) {
+            $error .= $text;
```
Comment (othercorey) [inline src/Http/Client/Auth/Oauth.php:384]: Looks ok, but would be nice to trigger one error in a test.

Comment (markstory) [reply src/Http/Client/Auth/Oauth.php:384]: Added a test for that.

**Pull request merged**